### DISCUSSION TITLE
Docs: Bun runtime support

### DISCUSSION THREAD

**Discussion by zigang93** [vercel/next.js#55272] [Ideas]
### What is the improvement or update you wish to see?

[bun](https://bun.sh/) is officially 1.0 and production ready.. I am looking forward to replace all test suit, package manager and  runtime in bun. 

I know that it may sound awkward to vercel teams that are working at turbopack.. but clearly bun is moving to good direction for js ecosystem

### Is there any context that might help us understand?

any possible to run bun runtime?

**Comment by panzacoder** [comment]
I have no expectations of immediate support, but I would definitely like to hear Vercel's take/stance, or even just to know if they have planned a timeframe to spike on bun support in the near future.

Thanks for opening the ticket!

**Comment by wwwdepot** [comment]
You can use Next.js with Bun, see PRs https://github.com/vercel/next.js/pull/53467  https://github.com/vercel/next.js/pull/53590.

Using Bun runtime on Vecel, not yet, see https://github.com/orgs/vercel/discussions/2021#discussioncomment-6967946

**Comment by panzacoder** [comment]
Well, kinda. According to Bun documentation, Bun can be used as a package manager, but Next uses Node APIs that are not compatible with Bun at the moment (annoyingly, there is no link to discussion around which APIs are blocking).

>Next.js currently relies on Node.js APIs that Bun does not yet implement. The guide below uses Bun to initialize a project and install dependencies, but it uses Node.js to run the dev server.
> [Source: Bun guide for Next.js](https://bun.sh/guides/ecosystem/nextjs) 

For me at least, there is little point in adopting bun if it doesn't replace all the relevant tooling. Even if there are edges, runtime support is the one that really matters for the majority of the benefits.

**Reply by jjhesk** [reply]
this is the same issue today.

**Comment by wwwdepot** [comment]
Tried example [app-dir-i18n-routing](https://github.com/vercel/next.js/tree/canary/examples/app-dir-i18n-routing) in my Ubuntu VM. Didn't encounter any error.

```
rm -rf yarn.lock node_modules

bun install

bun run build

bun run start
```

**Reply by andrewslock** [reply]
You probably didnt get any error because the command `bun run` uses nodejs by default (https://bun.sh/docs/cli/run#bun). If you want to run it using the bun runtime you need to set the `--bun` flag. So instead you should've tried `bun run --bun build` and `bun run --bun start`

**Reply by vincerubinetti** [reply]
Oh wow I had no idea this was a thing. I've been using Bun on several projects over the last few months, and it's been working great, but I wonder if I've actually still been running most things in Node because I have both Bun and Node installed.

**Comment by zigang93** [comment]
> You can use Next.js with Bun, see PRs #53467 #53590.
> 
> Using Bun runtime on Vecel, not yet, see https://github.com/orgs/vercel/discussions/2021#discussioncomment-6967946

Well, I am using docker in nextjs..
there are same issue that been reported in https://github.com/oven-sh/bun/issues/4795 ..

**Comment by christophstach** [comment]
I tried using bun with Next.js. Had no problem installing it, but I had to install the types manually. It wasn't installed automatically by the `bunx create-next-app` so I had to `bun install -d bun-types`. Also, I realized that I was not able to get the import working. I am always getting the following error:


```tsx
import Database from 'bun:sqlite'

const db = new Database("bun.db")
```

```bash
bun:sqlite
Module build failed: UnhandledSchemeError: Reading from "bun:sqlite" is not handled by plugins (Unhandled scheme).
Webpack supports "data:" and "file:" URIs by default.
You may need an additional plugin to handle "bun:" URIs.
```

Any news on that or how to solve it?

**Reply by ecumene** [reply]
https://github.com/oven-sh/bun/issues/4350#issuecomment-1697957252

**Comment by fungilation** [comment]
I'm tracking esbuild related issue (not running postinstall) in https://github.com/oven-sh/bun/issues/5148, which is blocking Vercel deploy of my app

**Comment by RobSchilderr** [comment]
I still had these errors when trying to run my [Turborepo starter project ](https://github.com/RobSchilderr/nextjs-native-starter) with Bun. If you go to the `bun` branch of that repo, you can see my changes and what I tried to achieve.

```
er.development.js:6141:7)
next-web:build: Error: <Html> should not be imported outside of pages/_document.
next-web:build: Read more: https://nextjs.org/docs/messages/no-document-import-in-page
```

and
```
next-web:build: Error occurred prerendering page "/". Read more: https://nextjs.org/docs/messages/prerender-error
next-web:build: Error: NextRouter was not mounted. https://nextjs.org/docs/messages/next-router-not-mounted
```



Steps to reproduce:

**1. Initial Setup:**
No package manager was set initially. Ran: `npx @turbo/codemod add-package-manager` and then `yarn global add @turbo/codemod`. This added a `packageManager` field to package.json.

**2. Switching to Bun:**
Updated package.json with: `"packageManager": "bun@1.0.2"`. Encountered a type error: `String does not match the pattern of "(npm|pnpm|yarn)@\d+\.\d+\.\d+(-.+)?".` This comes from the IDE so we can safely ignore for now.

**3. Cleaning Up:**
Removed `node_modules`, Ran `bun install`. Deleted yarn.lock and now have a `bun.lockb`.

**4. Building with Bun:**
Instead of yarn build, used `bun run turbo build`. Note: There's a difference in command syntax which caused a brief confusion.


**5. Result:**
The errors as shown above.

**Reply by shaneosullivan** [reply]
Is there any update here? Presumably any resolution is reliant on Bun becoming more compatible with Node so that it "just works", is there any record of ongoing work or issues tracking this?

**Comment by Christophvh** [comment]
Was wondering if someone from Vercel ever addressed this? Bun is currently much more mature, getting close to 100% NodeJS compatibility.

**Comment by Plopix** [comment]
I would love to see Bun Runtime directly on Vercel!

**Comment by PrinOrange** [comment]
I think it's time to add Bun support as now.
Bun has better performance and less memory using that cuts down over 70%, with good compatibility.

**Comment by tjazsilovsek** [comment]
What is the status here?

**Comment by nicholasoxford** [comment]
Would love to see this 🙏

**Comment by shahbazshueb** [comment]
Would love to see this 🙏 (1)

**Comment by johnny-woodtke** [comment]
I built a custom Bun runtime for Vercel serverless functions here: https://github.com/johnny-woodtke/vercel-bun

All it requires is a few lines in your project's `vercel.json`, and it will integrate perfectly with Vercel's GitHub webhook deployments alongside a Next.js app.

Example deployment here: https://vercel-bun-nine.vercel.app
- `/api`: https://vercel-bun-nine.vercel.app/api
- `/api/hello`: https://vercel-bun-nine.vercel.app/api/hello?firstName=Johnny&lastName=Woodtke

**Reply by jjhesk** [reply]
does it run?

**Reply by johnny-woodtke** [reply]
Yes it runs. However, right now, bun can only be used for API routes.

**Comment by johnny-woodtke** [comment]
To be honest, even though bun has come a long way, I think it's going to require the bun team to release an LTS version for Vercel/Next.js to natively support the bun runtime.

You can see in the [AWS Lambda docs](https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html#runtimes-future) that the AWS team only supports new managed runtimes when the given runtime releases an LTS version. You may also notice that all of [Vercel's officially supported runtimes](https://vercel.com/docs/functions/runtimes) are a subset of AWS Lambda's managed runtimes, which makes sense considering Vercel's serverless infrastructure is AWS Lambda under the hood.

I reckon it's pretty high risk for Vercel to endorse a runtime as production-ready by labeling it as an official runtime, so until bun releases an LTS version and AWS Lambda supports it natively, I think us bun devs are SOL.

**Reply by quantizor** [reply]
It wouldn't be the default, opt-in

**Reply by johnny-woodtke** [reply]
Correct, all of Vercel's non-Node runtimes are opt-in.

**Comment by williamrjribeiro** [comment]
It feels so close now: https://x.com/rauchg/status/1976708207528886333

**Comment by mybazzars** [comment]
Info.mybazzar@gmail.com

**Comment by alexandretrotel** [comment]
Coming soon I guess, really excited for this one!

**Comment by Christophvh** [comment]
https://vercel.com/blog/bun-runtime-on-vercel-functions

**Comment by jperezrealini** [comment]
I'm using the Bun runtime in Vercel and getting these errors sporadically:
```
Vercel Runtime Timeout Error: Task timed out after 300 seconds
```
The requests respond normally with the correct status, but the functions sometimes keep running until they time out 🤔
